x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -349,46 +349,23 @@ public void authorizeIndexAction(
-                // '-*' matches no indices so we allow the request to go through, which will yield an empty response
-                if (resolvedIndices.isNoIndicesPlaceholder()) {
-                    // check action name
-                    listener.onResponse(authorizeIndexActionName(action, authorizationInfo, IndicesAccessControl.ALLOW_NO_INDICES));
-                } else {
-                    assert resolvedIndices.getLocal().stream().noneMatch(Regex::isSimpleMatchPattern)
-                        || replaceable.indicesOptions().expandWildcardExpressions() == false
-                        || (request instanceof AliasesRequest aliasesRequest && aliasesRequest.expandAliasesWildcards() == false)
-                        : "expanded wildcards for local indices OR the request should not expand wildcards at all";
-                    listener.onResponse(
-                        buildIndicesAccessControl(
-                            action,
-                            authorizationInfo,
-                            Sets.newHashSet(resolvedIndices.getLocal()),
-                            aliasOrIndexLookup
-                        )
-                    );
-                }
-            }, listener::onFailure));
         } else {
             try {
-                final IndexAuthorizationResult indexAuthorizationResult = authorizeIndexActionName(
-                    action,
-                    authorizationInfo,
-                    IndicesAccessControl.ALLOW_NO_INDICES
-                );
-                if (indexAuthorizationResult.isGranted()) {
+                if (request instanceof IndicesRequest.Replaceable replaceable && replaceable.allowsRemoteIndices()
+                    || ensureRBAC(authorizationInfo).getRole().checkIndicesAction(action)) {
                     indicesAsyncSupplier.getAsync(ActionListener.wrap(resolvedIndices -> {
                         assert resolvedIndices.isEmpty() == false
                             : "every indices request needs to have its indices set thus the resolved indices must not be empty";
                         // all wildcard expressions have been resolved and only the security plugin could have set '-*' here.
                         // '-*' matches no indices so we allow the request to go through, which will yield an empty response
                         if (resolvedIndices.isNoIndicesPlaceholder()) {
-                            listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.ALLOW_NO_INDICES));
+                            if (request instanceof IndicesRequest.Replaceable replaceable
+                                && replaceable.allowsRemoteIndices()
+                                && ensureRBAC(authorizationInfo).getRole().checkIndicesAction(action) == false) {
+                                listener.onResponse(new IndexAuthorizationResult(true, IndicesAccessControl.DENIED));
```

Comment:
This is not really this PR. But I find the first parameter to `IndexAuthorizationResult` always confuses me, i.e. `true` then 2nd parameter is `DENIED`. In practice, the first parameter (`auditable`) is always `true` (the same goes for the superclass `AuthorizationResult`). This feels like a tech debt to me and we really should get rid of it. What do you think? I can raise a PR for it if you are happy with the proposal.

Reply:
Yes! I have it on my mind, and Slobodan also aluded to it.

> This feels like a tech debt to me and we really should get rid of it. What do you think? I can raise a PR for it if you are happy with the proposal.

Yes please, please raise the PR, that would be very appreciated!

Reply:
I raised https://github.com/elastic/elasticsearch/pull/91261

Please do not worry about merge conflicts. You can just go ahead and make changes as needed for this PR. I'll adapt mine after yours is merged. Thanks!